### PR TITLE
cli: added startproject command

### LINKED ISSUE #28: cli: add `instance create` command
This task should be built on top of PR #27 (which we can hopefully integrate later today)

Add a command ``startproject`` (name TBC) to ``inveniomanage`` that will generate an initial project structure using cookiecutter

Example of usage:
```inveniomanage startproject mysite```

will generate something a project structure like this:
https://github.com/lnielsen/mysite

The project structure should be generated using [cookiecutter Pythonic API](https://cookiecutter.readthedocs.org/en/latest/advanced_usage.html#calling-cookiecutter-functions-from-python). 

The cookiecutter template should be included in invenio-base and full path to found using ``pkg_resources``.

### PR DESCRIPTION
**Pull request by JavierDelgadoFernandez**
* Added startproject command which creates an skeleton
  of a new invenio site. (Closes #28)

Signed-off-by: Javier Delgado <javier.delgado.fernandez@cern.ch>

### CODE REVIEW

File: invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/celery.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+"""{{ cookiecutter.site_name }} base Invenio configuration."""
+
+from __future__ import absolute_import, print_function
+
+from flask_celeryext import create_celery_app
+
+from .factory import create_app
+
+# TODO: Import tasks
```
Comment (lnielsen) [inline invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/celery.py]: Just remove the todo.

File: invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/config.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+# Default language and timezone
+BABEL_DEFAULT_LANGUAGE = 'en'
+BABEL_DEFAULT_TIMEZONE = 'Europe/Zurich'
+I18N_LANGUAGES = [
+]
+
+BASE_TEMPLATE = "invenio_theme/page.html"
+
+SECRET_KEY = "sdfghsdfghfdghdfghjdfhjdfghjjhg"
```
Comment (lnielsen) [inline invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/config.py]: The secret key needs to be generated - i.e. here use:

```
SECRET_KEY = "{{cookiecutter.secret_key }}"
```

and in ``cmd.py`` and generation of a secret key (see https://github.com/inveniosoftware/invenio-ext/blob/master/invenio_ext/script/__init__.py#L79-L88 for how to generate a secret key)

File: invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# -*- coding: utf-8 -*-
+
+"""Mysite Invenio API application."""
+
+from invenio_base.app import create_app_factory
+from invenio_base.wsgi import create_wsgi_factory
+from invenio_config import default_conf_loader
```
Comment (lnielsen) [inline invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py]: ``from invenio_config import create_conf_loader``

**Comment by JavierDelgadoFernandez** [comment]
@lnielsen  Done.

File: invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+from . import config
+
+env_prefix = 'APP'
+
```
Comment (lnielsen) [inline invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py]:
```python
conf_loader = create_conf_loader(config=config, env_prefix=env_prefix)
```

File: invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+create_api = create_app_factory(
+    '{{ cookiecutter.site_name }}-api',
+    env_prefix=env_prefix,
+    config_loader=default_conf_loader(config, env_prefix=env_prefix),
```
Comment (lnielsen) [inline invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py]: ``conf_loader=conf_loader

File: invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    bp_entry_point='invenio_base.api_blueprints',
+    ext_entry_point='invenio_base.api_apps',
+)
+
+
+create_app = create_app_factory(
+    '{{ cookiecutter.site_name }}-app',
+    env_prefix=env_prefix,
+    config_loader=default_conf_loader(config, env_prefix=env_prefix),
```
Comment (lnielsen) [inline invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py]: ``conf_loader=conf_loader``

File: invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    '{{ cookiecutter.site_name }}-api',
+    env_prefix=env_prefix,
+    config_loader=default_conf_loader(config, env_prefix=env_prefix),
+    bp_entry_point='invenio_base.api_blueprints',
+    ext_entry_point='invenio_base.api_apps',
+)
+
+
+create_app = create_app_factory(
+    '{{ cookiecutter.site_name }}-app',
```
Comment (lnielsen) [inline invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py]: ``'{{ cookiecutter.site_name }}-app'`` -> ``'{{ cookiecutter.site_name }}'``

File: invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+from invenio_base.app import create_app_factory
+from invenio_base.wsgi import create_wsgi_factory
+from invenio_config import default_conf_loader
+
+from . import config
+
+env_prefix = 'APP'
+
+create_api = create_app_factory(
+    '{{ cookiecutter.site_name }}-api',
```
Comment (lnielsen) [inline invenio_base/cookiecutter-invenio-base/{{ cookiecutter.site_name }}/{{ cookiecutter.package_name }}/factory.py]: ``'{{ cookiecutter.site_name }}-api'`` -> ``'{{ cookiecutter.site_name }}'``

File: invenio_base/cmd.py [hunk 9]
```diff
@@ -23,14 +23,22 @@
...
 @click.command()
 @click.argument('name')
 def startproject(name):
     """Create a new project from template."""
-    # TODO: Use cookiecutter to create project template
+    path = resource_filename(__name__, "cookiecutter-invenio-base")
+
+    result = cookiecutter(path, no_input=True,
+                          extra_context={"site_name": name})
     click.echo("Creating project...")
```
Comment (lnielsen) [inline invenio_base/cmd.py]: ``click.secho("Created project...", fg="green")``

File: invenio_base/cmd.py [hunk 10]
```diff
@@ -23,14 +23,22 @@
...
 """Application bootstraping."""
 
-from __future__ import absolute_import, print_function, unicode_literals
+from __future__ import absolute_import, print_function
+
+from pkg_resources import resource_filename
 
 import click
 
+from cookiecutter.main import cookiecutter
```
Comment (lnielsen) [inline invenio_base/cmd.py]: Please check import sorting here using isort (see run-tests.sh) - not sure it's correct.

**Comment by lnielsen** [comment]
We nede to add a basic test of the generated cookiecutter template to Travis so that we're sure that it infects run.

Probably you'll need to add ``flask-celeryext`` and ``invenio-confg`` to test_requires in setup.py. Probably even just executing ``python manage.py`` will detect basic errors as it will create the flask application object.

**Comment by JavierDelgadoFernandez** [comment]
@lnielsen I am checking the a generated cookiecuter template using this:
```python
    with runner.isolated_filesystem():
        site_name = 'mysite2'
        result = runner.invoke(startproject, [site_name])
        path_to_folder = os.path.join(os.getcwd(), site_name)
        path_to_manage = os.path.join(path_to_folder, 'manage.py')
        assert call(['python',  path_to_manage]) == 0
```
To calculate the path I am using ```os.getcwd()``` which is [the same method used in the funcion CliRunner::isolated_filesystem](https://github.com/mitsuhiko/click/blob/master/click/testing.py#L305-L319).

**Pull request merged**